src/sql/mir.rs

```diff
@@ -1041,39 +1041,55 @@ impl SqlToMirConverter {
                 };
 
                 for &(&(ref src, ref dst), edge) in &sorted_edges {
-                    let jn = match *edge {
+                    let mut jns = Vec::new();
+                    match *edge {
                         // Edge represents a LEFT JOIN
                         QueryGraphEdge::LeftJoin(ref jps) => {
                             let (left_node, right_node) =
                                 pick_join_columns(src, dst, prev_node, &joined_tables);
-                            self.make_join_node(
-                                &format!("q_{:x}_n{}", qg.signature().hash, new_node_count),
-                                jps,
-                                left_node,
-                                right_node,
-                                JoinType::Left,
-                            )
+
+                            let mut prev_join = right_node;
+                            for jp in jps.into_iter() {
+                                let cur_join = self.make_join_node(
+                                    &format!("q_{:x}_n{}", qg.signature().hash, new_node_count),
+                                    jp,
+                                    left_node.clone(),
+                                    prev_join.clone(),
+                                    JoinType::Left,
+                                );
+
+                                prev_join = cur_join.clone();
+                                new_node_count+=1;
+                                jns.push(cur_join);
+                            }
                         }
                         // Edge represents a JOIN
                         QueryGraphEdge::Join(ref jps) => {
                             let (left_node, right_node) =
                                 pick_join_columns(src, dst, prev_node, &joined_tables);
-                            self.make_join_node(
-                                &format!("q_{:x}_n{}", qg.signature().hash, new_node_count),
-                                jps,
-                                left_node,
-                                right_node,
-                                JoinType::Inner,
-                            )
+
+                            let mut prev_join = right_node;
+                            for jp in jps.into_iter() {
+                                let cur_join = self.make_join_node(
+                                    &format!("q_{:x}_n{}", qg.signature().hash, new_node_count),
+                                    jp,
+                                    left_node.clone(),
+                                    prev_join.clone(),
+                                    JoinType::Inner,
+                                );
+
+                                prev_join = cur_join.clone();
+                                new_node_count+=1;
+                                jns.push(cur_join);
+                            }
```

Comment:
This code is identical to the one in the `LeftJoin` case, I think? It would be nice not to duplicate it, if it can easily be externalized into a closure. Makes it easier to avoid bugs that would occur if we change one of them but not the other.